lgc/patch/PatchBufferOp.cpp

```diff
@@ -64,129 +79,285 @@ PreservedAnalyses PatchBufferOp::run(Function &function, FunctionAnalysisManager
+                                  .add(&BufferOpLowering::postVisitLoadInst)
+                                  .add(&BufferOpLowering::postVisitMemCpyInst)
+                                  .add(&BufferOpLowering::postVisitMemSetInst)
+                                  .add(&BufferOpLowering::postVisitStoreInst)
+                                  .build();
+
+  SmallVector<Instruction *> instructions;
+  std::swap(instructions, m_postVisitInsts);
+  for (Instruction *inst : instructions)
+    visitor.visit(*this, *inst);
+  assert(m_postVisitInsts.empty());
+}
 
-  fixIncompletePhis();
+// =====================================================================================================================
+// Get the descriptor info describing whether the descriptor is invariant and/or divergent.
+//
+// This function resolves invariance and divergence to a "known" state if necessary.
+//
+// @param desc : the descriptor
+BufferOpLowering::DescriptorInfo BufferOpLowering::getDescriptorInfo(Value *desc) {
+  DescriptorInfo di = m_descriptors[desc];
+  if (di.invariant.has_value() && di.divergent.has_value())
+    return di;
+
+  // Resolve by first finding all necessary roots and then performing an aggressive data flow fixed point iteration,
+  // i.e. start with the tacit assumption that all descriptors are uniform and invariant.
+  DenseSet<Value *> seen;
+  SmallVector<Value *> searchWorklist;
+  SmallVector<Value *> propagationWorklist;
+  Value *current = desc;
+  for (;;) {
+    if (seen.insert(current).second) {
+      auto &di = m_descriptors[current];
+
+      if (!di.invariant.has_value() || !di.divergent.has_value()) {
+        if (auto *phi = dyn_cast<PHINode>(current)) {
+          for (Value *incoming : phi->incoming_values())
+            searchWorklist.push_back(incoming);
+        } else if (auto *select = dyn_cast<SelectInst>(current)) {
+          searchWorklist.push_back(select->getTrueValue());
+          searchWorklist.push_back(select->getFalseValue());
+        } else {
+          // Make conservative assumptions for unhandled instructions.
+          bool isConstant = isa<Constant>(current);
+          auto &di = m_descriptors[current];
+          if (!di.invariant.has_value())
+            di.invariant = isConstant;
+          if (!di.divergent.has_value()) {
+            // TODO: This would be entirely unnecessary if we had updatable divergence info.
+            di.divergent = !isConstant;
+          }
+        }
+      }
 
-  // Some instructions can modify the CFG and thus have to be performed after the normal visitors.
-  for (Instruction *const inst : m_postVisitInsts) {
-    if (MemSetInst *const memSet = dyn_cast<MemSetInst>(inst))
-      postVisitMemSetInst(*memSet);
-    else if (MemCpyInst *const memCpy = dyn_cast<MemCpyInst>(inst))
-      postVisitMemCpyInst(*memCpy);
+      if (!di.invariant.value_or(true) || di.divergent.value_or(false))
+        propagationWorklist.push_back(current);
```

Comment:
If I read correctly, I think the first loop doing backward traversal on data flow graph was trying to find descriptor which is divergent, right? If we did not find a divergent descriptor, the propagation loop will not be triggered. If that is true, then for a select with divergent condition and two uniform descriptor inputs. Then the forward propagation loop will not be triggered, right? It sounds like we should stop at such kind of `divergent select` and start forward propagation from it.

Reply:
You're right. I'm going to move the handling to the backwards loop.

The backwards search may still have to continue from the select because its invariance may not yet be known.